notification-service/src/main/java/com/tracking_delivery_system/notification_service/service/KafkaConsumer.java

```diff
@@ -0,0 +1,66 @@
+package com.tracking_delivery_system.notification_service.service;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.JsonMappingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.tracking_delivery_system.notification_service.model.LocationUpdate;
+import com.tracking_delivery_system.notification_service.model.StatusUpdate;
+
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.kafka.annotation.KafkaListener;
+import org.springframework.stereotype.Service;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+@Service
+@Slf4j
+public class KafkaConsumer {
+    private final Map<String, Map<String, Object>> deliveriesMap = new ConcurrentHashMap<>();
+    private final ObjectMapper mapper = new ObjectMapper();
+
+    @KafkaListener(topics = "location-update", groupId = "notification-service")
+    public void consumeTrackingUpdate(String trackingUpdate) {
+        try {
+            LocationUpdate locationUpdate = mapper.readValue(trackingUpdate, LocationUpdate.class);
+            String deliveryId = locationUpdate.getId().toString();
+            
+            deliveriesMap.computeIfAbsent(deliveryId, id -> new HashMap<>());
+            deliveriesMap.get(deliveryId).put("location-update", locationUpdate);
+
+            checkAndPrintMessage(deliveryId);
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @KafkaListener(topics = "delivery-status", groupId = "notification-group")
+    public void consumeDeliveryStatusUpdate(String statusUpdate) {
+        try {
+            StatusUpdate status = mapper.readValue(statusUpdate, StatusUpdate.class);
+
+            String deliveryId = status.getId().toString();
+            deliveriesMap.computeIfAbsent(deliveryId, id -> new HashMap<>());
+            deliveriesMap.get(deliveryId).put("delivery-status", status);
+
+            checkAndPrintMessage(deliveryId);
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException(e);
+        }
+
+    }
+
+    private synchronized void checkAndPrintMessage(String deliveryId) {
+        Map<String, Object> deliveryMap = deliveriesMap.get(deliveryId);
+
+        if (deliveryMap.containsKey("location-update") && deliveryMap.containsKey("delivery-status")) {
```

Comment:
i would recommend you divide this logic to have separate "if" for location-update and delivery-status because it is separate topics and and as i seen they are not combined. Or maybe i missed something - please correct me

Reply:
I used only one "if" for both topics, because inside this "if" I’m printing the messages from both topics in a single message. So this "if" ensures that the message will only be shown in the console when there is data from both topics.

Reply:
> i would recommend you divide this logic to have separate "if" for location-update and delivery-status because it is separate topics and and as i seen they are not combined. Or maybe i missed something - please correct me

I used only one "if" for both topics, because inside this "if" I’m printing the messages from both topics in a single message. So this "if" ensures that the message will only be shown in the console when there is data from both topics.